Tests

martink-rsa opened the pull request:
Writing tests for the app, currently in progress.

martink-rsa committed:
Tests: Basic url tests added

martink-rsa committed:
Tests: Testing issue with input updating

martink-rsa committed:
Test data

martink-rsa committed:
Tests: Issues

martink-rsa committed:
Tests: Fixed axios.get bug

martink-rsa committed:
Tests: SearchPage tests save point

martink-rsa committed:
Tests: Posts table added

martink-rsa committed:
Tests: Save point, adding beforeEach

martink-rsa committed:
Tests: Added input tests

martink-rsa committed:
Tests: Tests done

martink-rsa commented:
I've written the following tests, not limited to:
1. Header/Footer: 
1.1 Header and footer loads
2. HomePage:
1.1 Elements render
1.2 Route to /search/ is working
3. SearchPage:
1. Loading spinner shows and disappears
2. Default subreddit is loaded
3. User-changed subreddit is loaded
4. Heatmap is loaded with all buttons, values are correct and correct colours are used.
5. Posts table is loaded and rows are created

jkettmann commented on src/components/App/App.test.js:
```diff
@@ -1,9 +1,38 @@
+const renderWithTheme = (component) => (
+  render(
+    <Theme>
+      <GlobalStyle />
```
Do you need the GlobalStyle here? Jest and react-testing-library shouldn't care about the styles I think

jkettmann commented on src/components/App/App.test.js:
```diff
@@ -1,9 +1,38 @@
+);
+
+function setup() {
+  const { getByTestId, getAllByTestId, queryByTestId } = renderWithTheme(
```
I would directly return the result of `renderWithTheme`

```
function setup() {
  return renderWithTheme(
```

Otherwise you have to touch this function every time you need another one of the queries

martink-rsa replied:
Interesting. Thank you. Deconstructing like this didn't occur to me and is far better as I don't have to keep adding/removing the exported functions

jkettmann commented on src/components/Header/Header.js:
```diff
@@ -21,13 +21,13 @@ const headerLinks = [
-      <nav>
-        {headerLinks.map((item) => (
-          <NavLink key={item.title} to={item.link} alt={item.title}>
+      <nav data-testid="navbar">
```
I think it would be enough to use `getByText` here instead of `getByTestId` here and below. [Here is a document of priorities given to the single queries](https://testing-library.com/docs/guide-which-query). It's recommended to try testing like a real user. Test IDs should be the last resort

martink-rsa replied:
I've been thinking of doing that, but I'm sitting with a mental dilemma.

I would have to test for the strings in the links in the header, but then these tests will fail if someone ever changes the contents of the links. I thought that tests shouldn't be written this way, something along the lines of "tests shouldn't break if someone changes the content of the site".

martink-rsa replied:
I haven't been able to make this change. I will handle the others and come back to it.

jkettmann replied:
That's a good point. At the same time the wordings don't change so often typically. And if somebody changed the text inside an important link or button the tests would catch it.

Honestly, I think there are advantages to both approaches. I'd just go for the recommended one because of the reasons above and because you don't need so many test IDs in your components

jkettmann commented on src/components/Hero/Hero.js:
```diff
@@ -19,14 +19,14 @@ function Hero() {
         subreddit.
       </Subtitle>
-      <CTA as={Link} to={`/search/${DEFAULT_SUBREDDIT}`}>
+      <CTA as={Link} to={`/${DEFAULT_PATH}/${DEFAULT_SUBREDDIT}`} data-testid="hero-button">
```
It's a good idea to extract the path into a config, but the name is not so descriptive. Maybe call it `SEARCH_PATH`. It might also be good to create a function that builds the path since the `/${DEFAULT_PATH}/${DEFAULT_SUBREDDIT}` is spread over multiple components now. Something like `getSearchPath` that could look like this

```
function getSearchPath(subreddit = DEFAULT_SUBREDDIT) {
  return `/${SEARCH_PATH}/${DEFAULT_SUBREDDIT}`
}
```

jkettmann commented on src/components/Hero/Hero.js:
```diff
@@ -19,14 +19,14 @@ function Hero() {
         {DEFAULT_SUBREDDIT}
       </Caption>
-      <Link to={`/search/${DEFAULT_SUBREDDIT}`}>
+      <Link to={`/${DEFAULT_PATH}/${DEFAULT_SUBREDDIT}`} data-testid="hero-img">
```
Here the testid is not needed as well

jkettmann commented on src/pages/HomePage/HomePage.test.js:
```diff
@@ -0,0 +1,65 @@
+
+function setup() {
+  const history = createMemoryHistory();
+  const { getByTestId, getAllByTestId, queryByTestId } = renderWithHistory(
```
Here also rather return everything

jkettmann commented on src/pages/SearchPage/Heatmap/PostsTable/PostsTable.js:
```diff
@@ -49,9 +49,15 @@ function PostsTable({ info }) {
           {sortByMinutes(info).map((item) => (
             <TBody key={`${item.title}-`}>
-              <TRow>
+              <TRow data-testid="heatmap-table-row">
```
Rather test the content of the table

jkettmann commented on src/pages/SearchPage/Heatmap/PostsTable/PostsTable.js:
```diff
@@ -37,7 +37,7 @@ function PostsTable({ info }) {
       <PostsTableContainer>
         <Heading>Posts</Heading>
-        <Table>
+        <Table data-testid="heatmap-table">
```
I don't think you need this test id if you test the content of the table. If the content is there then the table is there as well

jkettmann commented on src/pages/SearchPage/SearchBar/SearchBar.js:
```diff
@@ -63,8 +65,9 @@ function SearchBar({ isLoading }) {
+          data-testid="search-input"
         />
-        <Button type="submit" inactive={!isValidSearch || isLoading}>
+        <Button type="submit" disabled={!isValidSearch || isLoading} data-testid="search-button">
```
Here the testid is not needed as well

jkettmann replied:
Well maybe it is because of the Search link in the header... Try if you can make it without, otherwise leave it in

jkettmann commented on src/pages/SearchPage/SearchPage.js:
```diff
@@ -9,14 +10,18 @@ import { DEFAULT_SUBREDDIT } from '../../config';
   const url = `https://api.pushshift.io/reddit/search/submission/?subreddit=${subreddit}&sort=desc&sort_type=score&after=${oneYearAgo}&size=500`;
-  const response = await fetch(url);
-  return response.json();
+  try {
```
This is basically a duplication of the try catch logic in the `handleSearch` function below. I'd remove it here. If you want to keep it here make sure to copy the error message over when throwing the error since it might be of interest for the calling function

```
throw new Error(error.message)
```

jkettmann commented on src/pages/SearchPage/SearchPage.js:
```diff
@@ -51,20 +56,21 @@ function SearchPage() {
       handleSearch(searchPath)
         .then((data) => {
-          if (data.data.length !== 0) {
+          if (data.length !== 0 && data !== 'Error') {
```
Does the API return `Error`? That is strange

jkettmann commented on src/pages/SearchPage/SearchPage.test.js:
```diff
@@ -0,0 +1,203 @@
+  test('Default input is used when search page is loaded', async () => {
+    const { getByTestId } = setup();
+    const input = getByTestId('search-input');
+    expect(input).toBeInTheDocument();
```
You don't need to add extra assertions `toBeInTheDocument` when you use `getBy*` or `findBy*`. Those will fail if the element is not there

jkettmann replied:
The same goes for other occurrences of `toBeInTheDocument` here

jkettmann replied:
I think this would still be worth refactoring. It's pretty obvious to someone who is experienced with testing that this can be simplified

jkettmann commented on src/pages/SearchPage/SearchPage.test.js:
```diff
@@ -0,0 +1,203 @@
+    fireEvent.click(button);
+    await waitForElementToBeRemoved(getByTestId('loading-spinner'));
+    expect(axiosMock.get).toHaveBeenCalledTimes(2);
+    expect(axiosMock.get).toBeCalledWith(expect.stringMatching(NEW_SUBREDDIT));
```
:+1:

jkettmann commented on src/pages/SearchPage/SearchPage.test.js:
```diff
@@ -0,0 +1,203 @@
+describe('Error and heatmap display states', () => {
+  test('Will show an error message when there are no results', async () => {
+    const { getByTestId } = setup();
+    expect(getByTestId('loading-spinner')).toBeInTheDocument();
+    await waitForElementToBeRemoved(getByTestId('loading-spinner'));
+    const error = getByTestId('error');
+    expect(error).toBeInTheDocument();
+    expect(error.textContent).toBe('0 results returned.');
```
Instead of these lines you could write

```
await findByText('0 results returned.');
```

With this assertion you test the spinner over and over again. It would be enough to do that once and only use `waitForElementToBeRemoved` if required

```
expect(getByTestId('loading-spinner')).toBeInTheDocument();
await waitForElementToBeRemoved(getByTestId('loading-spinner'));
```

jkettmann commented on src/pages/SearchPage/SearchPage.test.js:
```diff
@@ -0,0 +1,203 @@
+    const { getByTestId } = setup('pass', dummyPosts);
+    expect(getByTestId('loading-spinner')).toBeInTheDocument();
+    await waitForElementToBeRemoved(getByTestId('loading-spinner'));
+    expect(getByTestId('heatmap')).toBeInTheDocument();
```
Here also `findByTestId` would be enough

jkettmann commented on src/pages/SearchPage/SearchPage.test.js:
```diff
@@ -0,0 +1,203 @@
+  });
+  test('Will show an error message when the API call fails', async () => {
+    const { getByTestId } = setup('fail');
+    expect(getByTestId('loading-spinner')).toBeInTheDocument();
+    await waitForElementToBeRemoved(getByTestId('loading-spinner'));
+    const error = getByTestId('error');
+    expect(error).toBeInTheDocument();
+    expect(error.textContent).toBe('There was an error processing your request.');
```
See comment above

jkettmann commented on src/pages/SearchPage/SearchPage.test.js:
```diff
@@ -0,0 +1,203 @@
+});
+
+describe('Heatmap values', () => {
+  test('Heatmap values matches dummy data', async () => {
```
Nice

jkettmann commented on src/pages/SearchPage/SearchPage.test.js:
```diff
@@ -0,0 +1,203 @@
+    expect(heatmapValues).toEqual(dummyHeatmapValues);
+    expect(axiosMock.get).toHaveBeenCalledTimes(1);
+  });
+  test('Heatmap elements use the correct theme colours', async () => {
```
Even nicer! 

Fyi: this might also be tested with a visual regression test with cypress, percy etc. Those basically take a screenshot and compare them to a previous version. But that's out of scope for this project IMO. Just good to know for you when you have interviews or get a job. Not a lot of companies have those and they can be valuable and easy to implement

jkettmann commented on package.json:
```diff
@@ -6,6 +6,9 @@
     "@testing-library/user-event": "^10.0.0",
+    "axios": "^0.19.2",
+    "history": "^4.10.1",
+    "jest-axe": "^3.4.0",
```
what is this used for?

jkettmann commented on package.json:
```diff
@@ -6,6 +6,9 @@
     "@testing-library/jest-dom": "^5.1.1",
     "@testing-library/react": "^10.0.1",
     "@testing-library/user-event": "^10.0.0",
```
The testing stuff should go to dev dependencies to not bloat the production build

martink-rsa replied:
`"@testing-library/jest-dom": "^5.1.1",` will throw a ESLint error, but the rest were moved.

jkettmann commented on package.json:
```diff
@@ -18,7 +21,7 @@
     "start": "react-scripts start",
     "build": "react-scripts build",
-    "test": "react-scripts test",
+    "test": "TZ=UTC react-scripts test --env=jest-environment-jsdom-sixteen",
```
Is `TZ=UTC` used to have a constant timezone? Is there documentation for that in the jest docs or so? Just curious

martink-rsa replied:
I found it while Googling for a solution for getting the timezone to always be the same. According to others, it works without problems, however, some say it might not work on Windows. I don't have a dev setup for Windows so I can't verify this. I wasn't able to source the docs for it, but I'm guessing it's part of Webpack.

https://github.com/nodejs/node/issues/4230

I will probably switch to a package in the future to emulate a timezone. Still thinking of a way to approach it because if I move to Moment, then I may as well rewrite my dates to use Moment instead of my own functions. In that case, it will be a refactor iteration.

jkettmann replied:
Great, if it's working then it's fine. I experienced in a lot of companies that they don't care so much about devs on Windows since almost everyone is using Mac anyway with a few outliers using Linux.

jkettmann reviewed (changes requested):
All in all excellent work. Very detailed tests covering every detail afaik. You can simplify some tests by making use of the async `findBy*` instead of `getBy*` queries.

martink-rsa committed:
Tests: Posts table issues

martink-rsa committed:
Tests: Checked data of table

martink-rsa committed:
Tests: Checked data of table

martink-rsa commented:
I've tried to get rid of IDs (getByIds etc.) such as trying to find a header and footer tag using `role` but wasn't able but couldn't get it to work. Looking into it I think it's meant more for Aria roles. 

`jest-axe` was used to do a wide accessibility check which seems to be great to find missing parameters such as alt tags, but it's not a complete solution as I noticed that there are still contrast levels that don't pass but aren't found by the test.

jkettmann commented:
Don't worry about the ids, just keep it in my for the next time. I'd still suggest you refactor the things I mentioned [here](https://github.com/ooloo-io/reddit-timer-martink-rsa/pull/11/files#r415764027).

martink-rsa commented:
> 
> 
> Don't worry about the ids, just keep it in my for the next time. I'd still suggest you refactor the things I mentioned [here](https://github.com/ooloo-io/reddit-timer-martink-rsa/pull/11/files#r415764027).

Not sure how I missed this, thank you. I will be changing it.

martink-rsa committed:
Tests: Refactor to remove redundancy of tests

martink-rsa commented:
I've refactored to remove a lot of the redundancy that I had e.g. assigning a variable to the element as well as checking if it was in the document. Now I only assign the variable.

I don't know if I've gone too concise though as I've also refactored some parts to not include a variable and use the query within the check e.g.

`expect(getByTestId('heatmap')).toBeInTheDocument();`

I feel there is not much further I can reduce but this could just be a limitation of my skills.

jkettmann commented on src/pages/SearchPage/SearchPage.test.js:
```diff
@@ -0,0 +1,210 @@
+function getPosts(data, day, hour) {
+  const currentTable = data[day][hour];
+  const posts = [];
+  for (let i = 0; i < currentTable.length; i += 1) {
```
You could use `.map` here

```
const posts = currentTable.map(({ author, title }) => ({
  ...
)};
```

martink-rsa replied:
Trying to `map` it but I'm not sure if it's possible with what I'm doing. I'm creating a 1d array from all of the table cells so I can test each one easily. The reason for doing this was because it became to complex to loop and test the data while it was in its complex object form. This is why there are multiple `push` statements in each loop:

```js
    posts.push(author);
    posts.push(title);
    posts.push(created);
    posts.push(score);
    posts.push(num_comments);
```

`currentTable`'s values are rows in the table which are 5 different key/value pairs. `map` is only going to give me one opportunity/return to deal with all of the key/values for that particular row so I would only be able to return an entire row instead of returning each value so that the data is in a flattened 1d array.

The only thing I can think of at the moment is to keep it as a row (2d array) from the `.map` array and then use `.flat` to flatten it before it's returned.

martink-rsa replied:
I think I might be able to do it with 2 maps, the 2nd looping through keys. Attempting this at the moment and then will do a comparison at the end.

jkettmann replied:
You're right, `map` won't work. I didn't read the code properly. You can use `reduce` though.

```
const posts = currentTable.map((tmpPosts, { author, title }) => ({
  return tmpPosts.concat(author, title, created, score, num_comments);
)}, []);
```

Didn't test the concat so be careful :-)

jkettmann replied:
Btw the name `posts` is a bit confusing here. That's why I expected it to be an array of post objects.

martink-rsa replied:
> 
> 
> Btw the name `posts` is a bit confusing here. That's why I expected it to be an array of post objects.

You're right. I changed it as soon as you did the review as I immediately saw it was the cause of confusion. It was originally for each post, but the code evolved quite a bit and naming got lost along the way. Thanks.

martink-rsa replied:
I don't think it's going to be possible to do it elegantly with array methods, namely for the following reason:

1. I can't map through all of the keys as I have keys in my object, such as the Reddit post URL, but I'm not testing for the Reddit URL in the test.

2. The test itself loops through the array and checks for the item's text to be in the document. Since the data is a 1d array with no keys/values, I can't separate out content, such as the URL, so that it's not being tested for. I can't switch back to passing an object to the test as that creates massive problems when it comes to testing and trying to loop through each of the objects.

What I've got is, but the test will break for reasons above. I have to flatten the solution below otherwise I have nested arrays:

```js
function getPosts(data, day, hour) {
  const currentTable = data[day][hour];
  const cells = currentTable.map((row) => {
    return Object.keys(row).map(key => {
      if (key === 'created') {
        return displayHHMM(row[key]);
      }
      return row[key].toString();
    });
  }).flat();
  return cells;
}
```

I hate giving up on a coding problem, but I'm starting to think that the way I did initially is going to be one of the more elegant solutions to handle it. 

I await your input. I don't mind going at it for as long as I need to get it right, just not sure if I'm going to find something that works better.

jkettmann replied:
Sorry, the code in the comment was wrong. I wanted to use `reduce` instead of `map`.

```
const posts = currentTable.reduce((tmpPosts, row) => ({
  // Convert date obj to "hh:mm" with am/pm suffixed
  const created = displayHHMM(row.created);
  // Can't test numeric values, need to convert to string
  const num_comments = row.num_comments.toString();
  const score = row.score.toString();
  return tmpPosts.concat(row.author, row.title, created, score, num_comments);
)}, []);
```

jkettmann replied:
This will give you a 1D array with exactly like the one in your code if I'm right

martink-rsa replied:
Amazing!! I tried it the first time (I knew to switch `map` with `reduce`) but couldn't figure out the syntax error it had.

Here it is in its glory with the syntax error fixed and variables renamed. Thank you so much!

```js
function getCellData(data, day, hour) {
  const currentTable = data[day][hour];
  const cells = currentTable.reduce((tmpCells, row) => {
    // Convert date obj to "hh:mm" with am/pm suffixed
    const created = displayHHMM(row.created);
    // Can't test numeric values, need to convert to string
    const num_comments = row.num_comments.toString();
    const score = row.score.toString();
    return tmpCells.concat(row.author, row.title, created, score, num_comments);
  }, []);
  return cells;
}
```

jkettmann commented on src/pages/SearchPage/SearchPage.test.js:
```diff
@@ -0,0 +1,210 @@
+  });
+  test('Heatmap displays the correct timezone message', async () => {
+    const { getByTestId } = setup('pass', dummyPosts);
+    await waitForElementToBeRemoved(getByTestId('loading-spinner'));
```
Since this assertion is repeated often: a nice way to prevent this kind of repetition is to create a function inside setup. 

```
const setup = () => {
  const queries = render(...)

  const waitForData = async() => {
    await waitForElementToBeRemoved(getByTestId('loading-spinner'));
  }
	
  return {
    ...queries,
    waitForData,
  }
}
```

and then use it here like

```
await waitForData();
```

This way it's a bit more descriptive what the assertion is doing.

You can also create functions that are more complex like `searchNewSubreddit` where you enter a new subreddit and  submit the form or so. Just as an example, not saying that you should do that here.

Here is a nice blog post [AHA testing](https://kentcdodds.com/blog/aha-testing)

jkettmann reviewed (approved):
Looks great. I added a comment. Change your code if you like. If not feel free to merge ;-)

martink-rsa committed:
Tests: Refactor table data cell fn and spinner test

jkettmann reviewed (approved):
:+1:
